### PR TITLE
Drop the support of synchronous execution

### LINKED ISSUE #531: Drop the support of synchronous execution
The current WebNN spec supports both [asynchronous](https://www.w3.org/TR/webnn/#api-mlcontext-async-execution) and [synchronous](https://www.w3.org/TR/webnn/#api-mlcontext-sync-execution) execution modes. In particular, the synchronous execution mode, including [`MLContext.computeSync()`](https://www.w3.org/TR/webnn/#dom-mlcontext-computesync), [`ML.createContextSync()`](https://www.w3.org/TR/webnn/#dom-ml-createcontextsync) and [`MLGraphBuilder.buildSync()`](https://www.w3.org/TR/webnn/#dom-mlgraphbuilder-buildsync) methods, were introduced (only available in dedicated worker) for easy integration with ML frameworks written in C++ and compiled to Wasm, for example ONNXRuntime WebNN EP (Execution Provider) used the sync execution before. 

Chromium [WebNN prototype](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/modules/ml/ml_context.idl) supports both execution modes for implementation feedback. Chrome team encouraged WG to check [whether sync APIs are really necessary before its launch](https://bugs.chromium.org/p/chromium/issues/detail?id=1488162).

Recently, ONNXRuntim WebNN EP experimented ([onnxruntime#19145](https://github.com/microsoft/onnxruntime/pull/19145)) the async execution mode and compared the performance with sync. For sync execution, ONNXRuntime runs WebNN EP in a dedicated worker and calls WebNN `computeSync()` method there, the JavaScript user code in main thread communicate (via `postMessage`) with WebNN EP in worker thread through ONNXRuntime [Wasm proxy](https://onnxruntime.ai/docs/api/js/interfaces/Env.WebAssemblyFlags.html#proxy). For async execution, ONNXRuntime runs WebNN EP in main thread and calls WebNN async `compute()` method through [asyncify](https://emscripten.org/docs/porting/asyncify.html#usage-with-embind). 

According to the [test result](https://github.com/microsoft/onnxruntime/pull/19145#issuecomment-1905088459) across 35 models (including CNNs & transformers), the model inference time difference of the two execution modes is minimum. Actually, for GPU, the async is even slightly faster than sync (async / sync 103% in average). While for CPU, the async is a bit slower than sync (async / sync 95% in average). It's because WebNN EP on CPU has less operators support currently (referring to [implementation status](https://webmachinelearning.github.io/webnn-status/)). For each non-supported op, the model inference will fallback to run Wasm op and return to compute next WebNN sub-graph. The more ops fallback, the more async compute call (more asyncify overhead).

With more ops being supported by WebNN CPU/XNNPACK backend, the ops fallback would be less that means less asyncify overhead. And with [JSPI](https://github.com/WebAssembly/js-promise-integration/blob/main/proposals/js-promise-integration/Overview.md) (JavaScript Promise Integration) [coming](https://groups.google.com/a/chromium.org/g/blink-dev/c/YNuYkzmxlOY/m/KKhowQd5BgAJ), the asyncify overhead hopefully would become even less. The performance of async execution mode is expected to be faster. 

With [onnxruntime#19145](https://github.com/microsoft/onnxruntime/pull/19145) merged, ONNXRuntime WebNN EP is now only using WebNN async execution mode and won't use sync execution mode anymore.

Based on this implementation experience, the proposal for WebNN spec is to remove the support of sync execution. That would help simplify the spec as well as the implementation. Wasm ML framework could use WebNN async methods via asyncify today and migrate to JSPI once it is available.

### PR DESCRIPTION
**Pull request by huningxin**
Remove the definition and algorithm steps for
- ML.createContextSync()
- MLGraphBuilder.buildSync()
- MLContext.computeSync()

This PR also removes the synchronous execution description from Programming Model section.

The previous algorithm steps of `MLGraphBuilder.build()` depends on `MLGraphBuilder.buildSync()`. The algorithm steps of `MLGraphBuilder.buildSync()` have issue tracked by #457. That is left to a separate PR. This PR just merges the algorithm steps of `MLGraphBuilder.buildSync()` into `MLGraphBuilder.build()` with some modifications:
- Remove "The permissions and context validity have been checked..." note regarding to #501.
- Change throwing exception to rejecting the promise with exception.

Fix #531

@wchao1115 @anssiko @inexorabletash , PTAL. Thanks!

***
<a href="https://pr-preview.s3.amazonaws.com/huningxin/webnn/pull/548.html" title="Last updated on Feb 15, 2024, 8:52 AM UTC (7fb3187)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/webmachinelearning/webnn/548/eb06ccf...huningxin:7fb3187.html" title="Last updated on Feb 15, 2024, 8:52 AM UTC (7fb3187)">Diff</a>

### CODE REVIEW

**Comment by inexorabletash** [comment]
The bots seem to be having issues, but I've locally verified that there are no Bikeshed warnings or errors.

File: index.bs [hunk 1]
```diff
@@ -1662,46 +1540,28 @@ Build a composed graph up to a given output operand into a computational graph,
...
-                1. If |operand| was created as an input by the underlying platform:
-                    1. If |operand|.{{MLOperand/[[name]]}}] is not unique for |graphImpl|, then [=exception/throw=] a {{TypeError}}.
-                    1. Add |operand|.{{MLOperand/[[descriptor]]}} to |graph|.{{MLGraph/[[inputDescriptors]]}}[|operand|.{{MLOperand/[[name]]}}].
-                1. If |operand| was created as a constant by the underlying platform:
-                    1. Implementations MAY preprocess and optimize the tensor data of |operand| for the underlying platform.
-                1. Register |operand|.{{MLOperand/[[operand]]}} in |graphImpl| as graph output.
-                1. Register |operand|.{{MLOperand/[[operator]]}} to |graphImpl|.
-    1. Return |graph|.
+    1. Return |promise| and run the following steps [=in parallel=]:
+        1. If |outputs| is empty, then [=reject=] |promise| with a "{{TypeError}}" {{DOMException}}.
```
Comment (inexorabletash) [inline index.bs:1545]: This phrasing change is incorrect - [TypeError](https://webidl.spec.whatwg.org/#exceptiondef-typeerror) is not a [DOMException](https://webidl.spec.whatwg.org/#idl-DOMException), it's a distinct JS/interface type. This should just read: "reject promise with a TypeError."

File: index.bs [hunk 2]
```diff
@@ -1662,46 +1540,28 @@ Build a composed graph up to a given output operand into a computational graph,
...
+        1. [=map/For each=] |name| &rarr; |operand| of |outputs|:
+            1. If |name| is empty, then [=reject=] |promise| with a "{{TypeError}}" {{DOMException}}.
```
Comment (inexorabletash) [inline index.bs:1547]: Same - should just be "reject promise with a TypeError"

File: index.bs [hunk 3]
```diff
@@ -1662,46 +1540,28 @@ Build a composed graph up to a given output operand into a computational graph,
...
+        1. If any of the following sub-steps fail, then [=reject=] |promise| with an "{{OperationError}}" {{DOMException}}.
+            1. Let |graph| be a new {{MLGraph}}:
+                1. Set |graph|.{{MLGraph/[[context]]}} to [=this=].{{MLGraphBuilder/[[context]]}}.
+            1. Make a request to the underlying platform to:
+                1. Connect |graph| to a new [=implementation-defined=] graph implementation |graphImpl| given |graph|.
+                1. Set |graph|.{{MLGraph/[[implementation]]}} to |graphImpl|.
+            1. Make a request to the underlying platform to initialize the graph:
+                1. [=map/For each=] |name| &rarr; |operand| of |outputs|:
+                    1. If [=validating MLOperand=] given |operand| and [=this=] returns false, then [=reject=] |promise| with a "{{TypeError}}" {{DOMException}}.
```
Comment (inexorabletash) [inline index.bs:1556]: Same - should just be "reject promise with a TypeError"

File: index.bs [hunk 4]
```diff
@@ -1662,46 +1540,28 @@ Build a composed graph up to a given output operand into a computational graph,
...
+                    1. If |operand| was created as an input by the underlying platform:
+                        1. If |operand|.{{MLOperand/[[name]]}} is not unique for |graphImpl|, then [=reject=] |promise| with a "{{TypeError}}" {{DOMException}}.
```
Comment (inexorabletash) [inline index.bs:1558]: Same - should just be "reject promise with a TypeError"

**Review by inexorabletash — CHANGES_REQUESTED** [review]

File: index.bs (hunk 1 above)
Comment (inexorabletash) [reply index.bs:1545]: In markdown:
```
[=reject=] |promise| with a {{TypeError}}.
```

**Comment by zolkis** [comment]
PR preview failed, just wanted to call attention to common usage of "execute graph" and other algorithms from the "Synchronous execution" section also in asynchronous execution. These should be moved or merged into the async execution steps. It's hard to see from the diff if that was the case.

File: index.bs (hunk 1 above)
Comment (huningxin) [reply index.bs]: Fixed.

File: index.bs (hunk 2 above)
Comment (huningxin) [reply index.bs]: Fixed.

File: index.bs (hunk 3 above)
Comment (huningxin) [reply index.bs]: Fixed.

File: index.bs (hunk 4 above)
Comment (huningxin) [reply index.bs]: Fixed.

**Review by inexorabletash — CHANGES_REQUESTED** [review]

File: index.bs [hunk 5]
```diff
@@ -1662,46 +1540,28 @@ Build a composed graph up to a given output operand into a computational graph,
...
-                1. If |operand| was created as an input by the underlying platform:
-                    1. If |operand|.{{MLOperand/[[name]]}}] is not unique for |graphImpl|, then [=exception/throw=] a {{TypeError}}.
-                    1. Add |operand|.{{MLOperand/[[descriptor]]}} to |graph|.{{MLGraph/[[inputDescriptors]]}}[|operand|.{{MLOperand/[[name]]}}].
-                1. If |operand| was created as a constant by the underlying platform:
-                    1. Implementations MAY preprocess and optimize the tensor data of |operand| for the underlying platform.
-                1. Register |operand|.{{MLOperand/[[operand]]}} in |graphImpl| as graph output.
-                1. Register |operand|.{{MLOperand/[[operator]]}} to |graphImpl|.
-    1. Return |graph|.
+    1. Return |promise| and run the following steps [=in parallel=]:
+        1. If |outputs| is empty, then [=reject=] |promise| with a {{TypeError}}.
```
[the remark this answers is not in the document]
Comment (huningxin) [reply index.bs]: Good catch! Fixed in latest commit. Thanks!

File: index.bs [hunk 6]
```diff
@@ -1662,46 +1540,28 @@ Build a composed graph up to a given output operand into a computational graph,
...
+        1. [=map/For each=] |name| &rarr; |operand| of |outputs|:
+            1. If |name| is empty, then [=reject=] |promise| with a {{TypeError}}.
```
[the remark this answers is not in the document]
Comment (huningxin) [reply index.bs]: Fixed.

File: index.bs [hunk 7]
```diff
@@ -1662,46 +1540,28 @@ Build a composed graph up to a given output operand into a computational graph,
...
+        1. If any of the following sub-steps fail, then [=reject=] |promise| with an "{{OperationError}}" {{DOMException}}.
+            1. Let |graph| be a new {{MLGraph}}:
+                1. Set |graph|.{{MLGraph/[[context]]}} to [=this=].{{MLGraphBuilder/[[context]]}}.
+            1. Make a request to the underlying platform to:
+                1. Connect |graph| to a new [=implementation-defined=] graph implementation |graphImpl| given |graph|.
+                1. Set |graph|.{{MLGraph/[[implementation]]}} to |graphImpl|.
+            1. Make a request to the underlying platform to initialize the graph:
+                1. [=map/For each=] |name| &rarr; |operand| of |outputs|:
+                    1. If [=validating MLOperand=] given |operand| and [=this=] returns false, then [=reject=] |promise| with a {{TypeError}}.
```
[the remark this answers is not in the document]
Comment (huningxin) [reply index.bs]: Fixed.

File: index.bs [hunk 8]
```diff
@@ -1662,46 +1540,28 @@ Build a composed graph up to a given output operand into a computational graph,
...
+                    1. If |operand| was created as an input by the underlying platform:
+                        1. If |operand|.{{MLOperand/[[name]]}} is not unique for |graphImpl|, then [=reject=] |promise| with a {{TypeError}}.
```
[the remark this answers is not in the document]
Comment (huningxin) [reply index.bs]: Fixed.

File: index.bs [hunk 9]
```diff
@@ -1662,46 +1540,28 @@ Build a composed graph up to a given output operand into a computational graph,
...
-                1. If |operand| was created as a constant by the underlying platform:
-                    1. Implementations MAY preprocess and optimize the tensor data of |operand| for the underlying platform.
-                1. Register |operand|.{{MLOperand/[[operand]]}} in |graphImpl| as graph output.
-                1. Register |operand|.{{MLOperand/[[operator]]}} to |graphImpl|.
-    1. Return |graph|.
+    1. Return |promise| and run the following steps [=in parallel=]:
+        1. If |outputs| is empty, then [=reject=] |promise| with a {{TypeError}}.
+        1. [=map/For each=] |name| &rarr; |operand| of |outputs|:
+            1. If |name| is empty, then [=reject=] |promise| with a {{TypeError}}.
+        1. If any of the following sub-steps fail, then [=reject=] |promise| with an "{{OperationError}}" {{DOMException}}.
```
[the remark this answers is not in the document]
Comment (huningxin) [reply index.bs]: Fixed.

**Review by anssiko — APPROVED** [review]
Thanks @huningxin for preparing this PR swiftly to allow this change get into the next spec milestone.

And thanks @inexorabletash for your careful review.

@wchao1115 do you want to signal your +1 for this synchronous execution removal?

**Review by inexorabletash — APPROVED** [review]

**Review by wchao1115 — APPROVED** [review]

**Comment by anssiko** [comment]
By merging, we're checking some boxes:

- [x] This removal gets into the next Candidate Rec Snapshot (CRS) as anticipated in the high-level status https://www.w3.org/TR/webnn/#sotd
- [x] This also addresses feedback from earlier TAG review https://github.com/w3ctag/design-reviews/issues/771#issuecomment-1332346304 that suggested to drop this feature. (I gave TAG an update on this removal in our delta review.)

**Pull request merged**